packages/compose/src/hooks/use-modified-style/index.js

```diff
@@ -0,0 +1,82 @@
+// @ts-nocheck
+/**
+ * WordPress dependencies
+ */
+import { useState, useMemo, useEffect } from '@wordpress/element';
+
+/**
+ * Pluck the modified styles by matching the modifier in the selector name.
+ * The style selectors should follow the BEM naming convetion e.g.
+ *    block__element--modifier
+ * The return styles are keyed by the block_element only.
+ *
+ * @param {Object} styles
+ * @param {string} modifier - the base modifier without the `--` prefix
+ *
+ * @return {Object} - Subset of styles based on the modifier
+ */
+function selectModifiedStyles( styles, modifier ) {
+	const modifierSelectors = Object.keys( styles ).filter( ( selector ) =>
+		selector.match( `--${ modifier }$` )
+	);
+
+	return modifierSelectors.reduce( ( modifiedStyles, modifierSelector ) => {
+		const blockElementSelector = modifierSelector.split( '--' )[ 0 ];
+		modifiedStyles[ blockElementSelector ] = styles[ modifierSelector ];
+		return modifiedStyles;
+	}, {} );
+}
+
+function isModifierEnabled( modifier, states ) {
+	return states[ modifier ]?.every( ( truthy ) => truthy );
+}
+
+function mergeStyles( styles, styleUpdates ) {
+	return Object.keys( styleUpdates ).reduce(
+		( mergedStyles, selector ) => ( {
+			...mergedStyles,
+			[ selector ]: {
+				...mergedStyles[ selector ],
+				...styleUpdates[ selector ],
+			},
+		} ),
+		styles
+	);
+}
+
+function useModifiedStyle( baseStyles, modifierStates ) {
+	const [ styles, setStyles ] = useState( baseStyles );
+	const modifiers = Object.keys( modifierStates );
+
+	const updateStyles = useMemo( () => {
+		const modifiedBaseStyles = ( modifier ) =>
+			selectModifiedStyles( baseStyles, modifier );
+
+		const modifiedStyles = modifiers.reduce(
+			( mergedStyles, modifier ) => ( {
+				...mergedStyles,
+				[ modifier ]: modifiedBaseStyles( modifier ),
+			} ),
+			{ ...baseStyles }
+		);
+		return ( updatedStyles, selector ) =>
+			mergeStyles( updatedStyles, modifiedStyles[ selector ] );
+	}, [] );
+
+	// Convert the modified states to string to satisfy the equality check.
+	const modifiedStatesString = Object.values( modifierStates ).toString();
```

Comment:
@mchowning this is close to that neat `JSON.stringify` trick you found. I'm a bit worried about how it's used here https://github.com/WordPress/gutenberg/commit/557ee25b389bda009bd1821cab6c3209915ba741 since `JSON.stringify` can be slow and the `baseStyles` could be quite large (also I really don't think this hook should be used for dynamic styles). Here we are just calling `toString` on an array which should be  fast and gives us something check equality against.  It will actually flatten out the values which is cool : 
```js 
[ [true, true] , false].toString() === "true,true,false"
[[true,true],[false]].toString() === "true,true,false"
```

Reply:
> It will actually flatten out the values

That is so interesting. I never would have guessed that. JS is so weird! 😆

>  I'm a bit worried about how it's used here 557ee25 since JSON.stringify can be slow and the baseStyles could be quite large (also I really don't think this hook should be used for dynamic styles).

👍  That's a totally valid concern, particularly if the baseStyles could get large. I'm fine with not using my `stringify-everything!!!` proposal. 🙂 

> Here we are just calling toString on an array which should be fast and gives us something check equality against. It will actually flatten out the values which is cool

Doesn't that give us some pretty surprising behavior depending on how the modifiers are changed?

```
const justActive = Object.values({active: true}).toString()
const justSpecial = Object.values({dark: true}).toString()
justActive === justSpecial
```

and similarly:

```
const activeLight = Object.values({active: true, dark: false}).toString()
const inactiveDark = Object.values({dark: true, active: false}).toString()
activeLight === inactiveDark
```

Would we be better off just stringifying the entire `modierStates` object?

Reply:
Ok , I see that issue. _ideally_ the modifier keys would be static along with the baseStyles but I don't think there is a clear way to enforce that. I agree then that checking the entire `modifierStates` object is more secure. It does mean we'll have to use `JSON.stringify` but these objects should likely be simple enough to go that route.